src/flint/types/fq_default.pyx

```diff
@@ -0,0 +1,790 @@
+            >>> int(gf(123))
+            123
+        """
+        cdef fmpz x = fmpz.__new__(fmpz)
+        res = fq_default_get_fmpz(x.val, self.val, self.ctx.val)
+        if res == 1:
+            return int(x)
+        raise ValueError("fq element has no lift to the integers")
+
+    def polynomial(self):
+        """
+        Returns a representative of ``self`` as a polynomial in `(Z/pZ)[x] / h(x)`
+        where `h(x)` is the defining polynomial of the finite field.
+
+            >>> gf = fq_default_ctx(163, 3)
+            >>> a = gf([1,2,3])
+            >>> a.polynomial()
+            3*x^2 + 2*x + 1
+            >>> gf(123).polynomial()
+            123
+        """
+        cdef fmpz_mod_poly pol
+
+        ring_ctx = fmpz_mod_poly_ctx(self.ctx.prime())
+        pol = ring_ctx.new_ctype_poly()
+        fq_default_get_fmpz_mod_poly((<fmpz_mod_poly>pol).val, self.val, self.ctx.val)
+
+        return pol
+
+    def str(self):
+        return self.polynomial().str(var=self.ctx.var.decode())
+
+    def __repr__(self):
+        # TODO: what do we want here?
+        return str(self)
+
+    # =================================================
+    # Comparisons
+    # =================================================
+    def is_zero(self):
+        """
+        Returns true is self is zero and false otherwise
+
+            >>> gf = fq_default_ctx(163, 3)
+            >>> gf(0).is_zero()
+            True
+            >>> gf(-1).is_zero()
+            False
+        """
+        return 1 == fq_default_is_zero(self.val, self.ctx.val)
+
+    def is_one(self):
+        """
+        Returns true is self is one and false otherwise
+
+            >>> gf = fq_default_ctx(163, 3)
+            >>> gf(-1).is_one()
+            False
+            >>> gf(1).is_one()
+            True
+        """
+        return 1 == fq_default_is_one(self.val, self.ctx.val)
+
+    def __richcmp__(self, other, int op):
+        cdef bint res
+        if op != 2 and op != 3:
+            raise TypeError("fq_default cannot be ordered")
+
+        if not typecheck(other, fq_default):
+            other = self.ctx.any_as_fq_default(other)
+
+        if typecheck(other, fq_default):
+            res = (self.ctx == (<fq_default>other).ctx) and \
+                  fq_default_equal(self.val, (<fq_default>other).val, self.ctx.val)
+            if op == 2:
+                return res
+            else:
+                return not res
+        else:
+            return NotImplemented
+
+    # =================================================
+    # Generic arithmetic required by flint_scalar
+    # =================================================
+
+    def _neg_(self):
+        cdef fq_default res
+        res = self.ctx.new_ctype_fq_default()
+        fq_default_neg(res.val, self.val, self.ctx.val)
+        return res
+
+    def _add_(self, other):
+        other = self.ctx.any_as_fq_default(other)
+        if other is NotImplemented:
+            return NotImplemented
```

Comment:
One of the things that would be nice about having the `__add__` etc boilerplate in the superclass is if we could centralise the logic around how coercions happen. Ideally if we have the overhead of an `__add__` method that calls an `_add_` method then the `__add__` method should do something to simplify the situation like handling NotImplemented etc. It would be like:
```python
class Super:
    def __add__(self, other):
        # Happy fast path:
        if self.typecheck(other):
            return self._add_(other)
        elif typecheck(other, int):
            return self._add_int_(other)
        other = self.any_as_self(other)
        if other is NotImplemented:
            return NotImplemented
        return self._add_(other)

class Sub:
    def any_as_self(self, other):
        ...

    def _add_(Sub self, Sub other):
        # No type checking needed here
        # No NotImplemented etc. Types must be correct.
        r = Sub.__new__(Sub)
        fmpsub_add(r.val, self.val, other.val)
        return r
```
For some types like fmpz this would be too much overhead and we would still need to have `__add__` implemented fully with all boilerplate in th subclass. The intention though should be that the superclass is somehow simplifying the situation to make it easier for the subclass to implement these methods. Otherwise there isn't any point in having `__add__` call `_add_`.

Reply:
I have tried to do something similar to this in the latest commit, although without the optimisation of things like `self._add_int_(other)` which I think we can do at a later date?